File: tool/create_dict_db.py
import sqlite3
import os

def main():
    db_path = 'assets/db/dictionary.sqlite'
    os.makedirs('assets/db', exist_ok=True)
    
    if os.path.exists(db_path):
        os.remove(db_path)
        
    print("Opening database...")
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()
    
    # Create table matching Drift schema (mostly)
    # Drift adds 'id' INTEGER PRIMARY KEY AUTOINCREMENT
    cursor.execute('''
        CREATE TABLE dictionary (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            word TEXT UNIQUE,
            definition TEXT
        )
    ''')
    
    # 1. Load Definitions
    print("Loading definitions...")
    defs_map = {}
    try:
        with open('assets/ODS9_definitions.txt', 'r', encoding='utf-8') as f:
            for line in f:
                line = line.strip()
                if not line: continue
                
                parts = line.split(None, 1) # Split on first whitespace
                if len(parts) < 2: continue
                
                word_part = parts[0]
                def_part = parts[1]
                
                # Clean word (remove variants like ",ENNE")
                if ',' in word_part:
                    word_part = word_part.split(',')[0]
                
                defs_map[word_part] = def_part
        print(f"Loaded {len(defs_map)} definitions.")
    except FileNotFoundError:
        print("Warning: Definitions file not found.")

    # 2. Load Words
    print("Loading words...")
    try:
        with open('assets/ODS9.txt', 'r', encoding='utf-8') as f:
            words = [line.strip().upper() for line in f if line.strip()]
    except FileNotFoundError:
        print("Error: ODS9.txt not found.")
        return

    print(f"Inserting {len(words)} words...")
    
    # Batch insert
    batch_size = 10000
    data = []
    
    for word in words:
        definition = defs_map.get(word) # None if not found
        data.append((word, definition))
        
        if len(data) >= batch_size:
            cursor.executemany('INSERT OR IGNORE INTO dictionary (word, definition) VALUES (?, ?)', data)
            data = []
            print(f"Processed...")
            
    if data:
        cursor.executemany('INSERT OR IGNORE INTO dictionary (word, definition) VALUES (?, ?)', data)
        
    conn.commit()
    conn.close()
    print("Done! Dictionary generated.")

if __name__ == "__main__":
    main()

File: tool/create_dict_db.dart
import 'dart:io';
import 'package:sqlite3/sqlite3.dart';

void main() {
  final dbPath = 'assets/db/dictionary.sqlite';
  // Clean old file
  if (File(dbPath).existsSync()) {
    File(dbPath).deleteSync();
  }
  
  // Ensure dir exists
  Directory('assets/db').createSync(recursive: true);

  print('Opening database...');
  final db = sqlite3.open(dbPath);
  
  // Create table
  db.execute('''
    CREATE TABLE dictionary (
      word TEXT PRIMARY KEY,
      definition TEXT
    );
  ''');

  // 1. Load Definitions (Map)
  print('Loading definitions...');
  final defsFile = File('assets/ODS9_definitions.txt');
  final defsMap = <String, String>{};
  
  if (defsFile.existsSync()) {
    final lines = defsFile.readAsLinesSync();
    for (final line in lines) {
      if (line.trim().isEmpty) continue;
      
      // Format: "MOT  Def..." or "MOT ,VAR  Def..."
      // On cherche le premier espace double ou espace simple après le mot ?
      // Le fichier a l'air d'utiliser des espaces.
      // Ex: "AA  Géol. n. Coulée..."
      
      final firstSpace = line.indexOf(' ');
      if (firstSpace == -1) continue;
      
      var wordPart = line.substring(0, firstSpace).trim();
      final defPart = line.substring(firstSpace).trim();
      
      // Nettoyage wordPart (ex: "AALENIEN ,ENNE")
      // On garde juste "AALENIEN" pour la clé principale
      if (wordPart.contains(',')) {
        wordPart = wordPart.split(',')[0].trim();
      }
      
      defsMap[wordPart] = defPart;
    }
    print('Loaded ${defsMap.length} definitions.');
  } else {
    print('Warning: Definitions file not found.');
  }

  // 2. Load Words and Insert
  print('Loading words and inserting...');
  final wordsFile = File('assets/ODS9.txt');
  if (!wordsFile.existsSync()) {
    print('Error: ODS9.txt not found.');
    return;
  }

  final words = wordsFile.readAsLinesSync();
  final stmt = db.prepare('INSERT OR IGNORE INTO dictionary (word, definition) VALUES (?, ?)');
  
  int count = 0;
  db.execute('BEGIN TRANSACTION');
  
  for (final line in words) {
    final word = line.trim().toUpperCase();
    if (word.isEmpty) continue;
    
    // Look for definition
    // Si c'est une forme fléchie, on n'a pas la def directe.
    // On pourrait essayer de trouver si le mot "commence par" une racine connue ?
    // Risqué (ABAISSAIT vs ABAISSER).
    // Pour l'instant, on met la def exacte si trouvée.
    final def = defsMap[word];
    
    stmt.execute([word, def]);
    count++;
    
    if (count % 10000 == 0) print('$count words processed...');
  }
  
  db.execute('COMMIT');
  stmt.dispose();
  db.dispose();
  
  print('Done! Created dictionary.sqlite with $count entries.');
}
